src/awkward/_backends/dispatch.py

```diff
@@ -129,3 +132,22 @@ def regularize_backend(backend: str | Backend) -> Backend:
         return _name_to_backend_cls[backend].instance()
     else:
         raise ValueError(f"No such backend {backend!r} exists.")
+
+
+@register_backend_lookup_factory
+def find_virtual_backend(obj: type):
+    """
+    Implements a lookup for finding the backends of virtual arrays.
+    This is necessary to avoid calling `isinstance` inside `backend_of_obj` which may cause slowdown.
+    """
+    if issubclass(obj, VirtualArray):
+
+        def finder(obj: VirtualArray):
+            if isinstance(obj.nplike, ak._nplikes.numpy.Numpy):
+                return _name_to_backend_cls["cpu"].instance()
+            elif isinstance(obj.nplike, ak._nplikes.cupy.Cupy):
+                return _name_to_backend_cls["cuda"].instance()
```

Comment:
This is much better. In future, I'd like to update `_backends` so that we don't have to hard-code the nplike-to-backend relationship.